Fix/notification

HyeonSeongIM opened the pull request:
[#15] 알림 API 생성

알림 list, send, read, read-all, delete, delete-all 구현하였습니다.

데이터쪽에서는 is_read 생성하여 알림을 읽음 여부 확인하기 위해 추가로 만들었습니다.

알림 read 시 감정분석 결과 url로 redirect 하게 만들었습니다.

HyeonSeongIM committed:
Entity 코드 생성
- ERD 설계에 맞게 작성
- 모두 단방향 매핑으로 설계

주석 위치 수정
- 각 세부 코드 위에 주석 옮겼습니다.

HyeonSeongIM committed:
Merge branch 'dev' of https://github.com/P-Diary/diary-be into dev

# Conflicts:
#	src/main/java/com/diary/myDiary/domain/diary/entity/Diary.java

HyeonSeongIM committed:
Entity 코드 생성
- ERD 설계에 맞게 작성
- 모두 단방향 매핑으로 설계

주석 위치 수정
- 각 세부 코드 위에 주석 옮겼습니다.

HyeonSeongIM committed:
Merge branch 'dev' of https://github.com/P-Diary/diary-be into dev

HyeonSeongIM committed:
merge

HyeonSeongIM committed:
Merge branch 'dev' of https://github.com/P-Diary/diary-be into dev

HyeonSeongIM committed:
Feat: Notification api

HyeonSeongIM committed:
Feat: Notification api

HyeonSeongIM committed:
Feat: Notification api

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/entity/Notification.java:
```diff
@@ -17,17 +15,35 @@
-    private Long id;
+    // NotificationId
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "notification_id")
+    private Long notificationId;
```
notification_id 로 변경한 이유는 알 것 같지만 모든 테이블에 id값은 pk로 두는 것이 보편적?이기 때문에 별도로
notification_id 등으로 변경할 필요는 없을 것 같습니다
또한 가독성 문제도 있을 거구용 (notification.id -> notification.notification_id)

HyeonSeongIM replied:
넵 확인했습니다!

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/entity/NotificationType.java:
```diff
@@ -5,7 +5,7 @@
 public enum NotificationType {
 
-    EMOTION_ANALYSIS("감정 분석 결과"),
+    EMOTION_ANALYSIS("감정 분석 결과가 나왔습니다.")
```
👍

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/entity/Notification.java:
```diff
@@ -17,17 +15,35 @@
     @Enumerated(EnumType.STRING)
     private NotificationType notificationType;
 
+    // 메시지
+    @Column(nullable = false)
     private String message;
 
-    private LocalDateTime readAt;
+    // 읽음을 확인 (y, n)
+    @Column(nullable = false)
+    private Boolean is_read;
```
개인적으로 읽는 시간을 저장하면 읽기 여부 또한 같이 판단할 수 있다고 생각하지만.. Boolean 타입으로 가져가는 이유가 있을까요?
또한 변경한다면 코드 컨벤션을 유지하면 좋을 것 같습니다 :) `is_read` -> `isRead`

HyeonSeongIM replied:
대부분 알림들은 생성된 시간은 표기 되지만 읽은 시간은 표기가 되지 않아 is_read로 다시 생성하였습니다.
Boolean 타입으로 가져간 이유는 읽음, 읽지 않음 둘로 명확하게 나뉘어 String으로 값을 넣는 것 보다 Boolean으로 넣는 것이 더 편하다고 생각하여 만들었습니다.

코드 컨벤션 유지하기 위해 isRead로 바꿨습니다!

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/entity/Notification.java:
```diff
@@ -17,17 +15,35 @@
+    @Column(nullable = false)
+    private Boolean is_read;
+
+    @CreatedDate
+    @Column(updatable = false)
+    private LocalDateTime createdAt;
```
`BaseTimeEntity` 가 상속된 부분을 확인해주세요
JPA에서 자주 사용되는 created, modified 에 대해 자동적으로 값을 주입해주는 친구입니다
JPA Auditing에 대해 검색하면 될 것 같아요

HyeonSeongIM replied:
@EnableJpaAuditing 부분 확인했습니다!

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/dto/NotificationDto.java:
우선 우리 프로젝트에선 Dto 관련해서 class 보단 record를 우선해서 사용하고 있습니다
record의 불변객체가 가지는 이점에 대해 검색해보시면 좋을 것 같아요
또한 코드 컨벤션을 유지하기 위해 Dtos 보단 Request, Response 등으로 명확한 클래스명을 나타내면 어디서 사용하는
Dto인지 좀 더 파악하기 편할 것 같네요

`@Data`가 가지는 어노테이션에 대해서도 간단하게 검색해보면 좋을 것 같네요
현재 파일에선 `@Getter`와 중복됩니다

HyeonSeongIM replied:
넵! 확인했습니다.

레코드의 장점 
1) 코드의 간결성 유지 -> 보일러플레이트 코드 대폭 줄일 수 있음
2) 불변성 -> 모든 필드는 final -> 안정성은 높이고 버그 발생 가능성 줄임

@Data
@ToString, @Getter, @Setter, @EqualsAndHashCode, @RequiredArgsConstructor 포함

@Data 사용을 지양하는 이유
Setter 남용 -> id 필드가 변경 허용하지 않으면 setter 도 없어야 안정성이 보장됨

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/service/NotificationService.java:
```diff
@@ -1,10 +1,26 @@
-    void sendNotification(Long memberId, String message);
-    void readNotification(Long notificationId);
+    // 알림 보내기
+    void sendNotification(NotificationDto notificationDto);
+
+    // 알림 리스트
+    List<NotificationDto> getNotification(Long memberId);
+
+    // 알림 읽기 후 리다이렉트
+    String readNotification(Long notificationId);
+
+    // 알림 전체 읽기
     void readAllNotification(Long memberId);
+
+    // 알림 삭제하기
     void deleteNotification(Long notificationId);
+
+    // 알림 전체 삭제하기
     void deleteAllNotification(Long memberId);
```
개인적으로 interface는 그 객체의 역할을 정하는 부분이라 생각합니다
이 객체가 가져야 할 역할에 대해 생각하고 필요한 매개변수는 무엇인지, 
반환해야 할 것은 어떤 것이 있는지에 대한 고민이 필요해 보입니다

sendNotification에서 Dtos를 입력받고 있는데 Dto에 내용을 살펴보면 notification_id, notification_type, isRead 등 
Dto가 파악하기 힘든 정보들을 같이 요청받고 있네요. Request에서 필요한 정보를 받을 때는 딱 필요한 정보만 받는게 좋을 것 같습니다

자주 언급하는 것 같지만 다른 도메인에서 코드 컨벤션을 확인하면 좋겠습니다 ㅎㅎ
대부분의 Response에서 명확한 Response 클래스를 정의하고 그 값을 반환하고 있네요

read메소드가 String, void 타입인 것은 아직 구현중이라 생각하겠습니당 🤔

메소드명에 대한 주석은 좋습니다 👍

HyeonSeongIM replied:
확인했습니다!

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/service/impl/NotificationServiceImpl.java:
```diff
@@ -1,38 +1,106 @@
     @Override
-    public void sendNotification(Long memberId, String message) {
+    public void sendNotification(NotificationDto notificationDto) {
+        Member member = memberRepository.findById(notificationDto.getMemberId())
+                .orElseThrow(() -> new RuntimeException("멤버 없음"));
```
Exception에 대한 부분은 global에서 관리되고 있습니다
각 도메인별로 ErrorCode에 대해 작성하고, globalExceptionHandler에 대해 공부해보시면 좋겠네요

또한 인터페이스에서 위의 exception같은 간단하고 중복적인 코드는
```java
    default Diary getByIdOrThrow(Long id) {
        return findById(id).orElseThrow(() -> new DiaryException(ErrorCode.NOT_FOUND_DIARY));
    }
```

처럼 정의하고 사용할 수 있습니다

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/service/impl/NotificationServiceImpl.java:
```diff
@@ -1,38 +1,106 @@
+        Member member = memberRepository.findById(notificationDto.getMemberId())
+                .orElseThrow(() -> new RuntimeException("멤버 없음"));
+
+        Notification notification = Notification.builder()
+                .member(member)
+                .notificationType(NotificationType.EMOTION_ANALYSIS)
+                .message(notificationDto.getMessage())
+                .is_read(false)
+                .build();
```
빌더 패턴을 사용한 것은 매우 좋습니다 😀
정적 패토리 메소드 패턴에 대해서도 공부하고 적용하면 좋겠습니다

HyeonSeongIM replied:
정적 팩토리 메소드 관련하여 공부하겠습니다!
조언 감사합니다!

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/service/impl/NotificationServiceImpl.java:
```diff
@@ -1,38 +1,106 @@
+    private final MemberRepository memberRepository;
 
+    // 알림 보내기
     @Override
-    public void sendNotification(Long memberId, String message) {
```
sendNotification 부분에는 `@Transactional` 이 필요 없는 걸까요??

HyeonSeongIM replied:
필요합니다! 확인하지 못하고 push 한 점 죄송합니다.

khyu2 commented on src/main/java/com/diary/myDiary/domain/notification/controller/NotificationController.java:
```diff
@@ -12,22 +15,44 @@ public class NotificationController {
-        notificationService.sendNotification(memberId, message);
+    // 알림 보내기
+    @PostMapping("/send")
+    @ResponseBody
```
`@ResponseBody`는 `@RestController`를 사용할 경우 하위 메소드는 전부 적용되는 것으로 알고 있는데
중복적으로 사용한 이유가 있을까요?

HyeonSeongIM replied:
지식이 얕았습니다. 수정하겠습니다!

khyu2 reviewed (commented):
잘 구현했지만 부족한 부분이 있어 보입니다 :)

HyeonSeongIM committed:
Feat: Notification api 수정 (1)
1) id 수정
2) Boolean 타입 comment
3) baseTimeEntity 내용 이해
4) dto -> record

HyeonSeongIM committed:
Feat: Notification api 수정 (2)
1) dto 삭제 후 record 생성
2) 정적 팩토리 메서드 사용
-> 오류 발생합니다.
Line:24 와 Line:34 확인해주세요.
3) Service 메서드 수정 완료
4) Controller에서 Operation 어노테이션 사용하였습니다.

HyeonSeongIM commented on src/main/java/com/diary/myDiary/domain/notification/service/impl/NotificationServiceImpl.java:
```diff
@@ -21,88 +20,68 @@ public class NotificationServiceImpl implements NotificationService {
-                .isRead(false)
-                .build();
-
+        Notification notification = Notification.from(member,notificationType, message, isRead);
```
여기서 from 사용할려는데 위에 24~25번째 줄 안해주면 오류 발생합니다.
notificationType는 import가 아예 안되고 isRead는 그냥 안됩니다.
확인 한번만 해주시면 감사합니다!

HyeonSeongIM committed:
Feat: Notification api 수정 (3)
예외처리 완료하였습니다.

HyeonSeongIM committed:
Merge 시도

HyeonSeongIM committed:
Merge 시도

HyeonSeongIM committed:
[#16] GPT에서 알림보내기

HyeonSeongIM committed:
[#16] GPT에서 알림보내기
